feat: add guards to session management

lamchau opened the pull request:
* added a fix/workaround for `rich` color handling 
![image](https://github.com/user-attachments/assets/03eeb10f-dfe7-4137-9c6d-67e2b1c6ebc2)
![image](https://github.com/user-attachments/assets/3ae161f0-8687-4f3a-a512-b89280b210bf)

 * switched `[dim]` to `[bold]` for dark terminals
![image](https://github.com/user-attachments/assets/d0e53840-27b9-48d1-8b84-6022eafcd1d4)

 * added a cleanup for empty sessions: edge case when a user aborts early - without it empty sessions are generated so we're unable to find the "right" last valid session to resume
 * added a prompt menu for overwrites: added a check for existing session files and give a user an opportunity to recover before they take a destructive action
![image](https://github.com/user-attachments/assets/e4aed727-eb2b-4a59-9223-33fd8a62568e)

note: the image shows both the empty session file (`goose session list` doesn't contain `abcd.jsonl`) and what the menu ux looks like for a session conflict

lifeizhou-ap commented:
Hi @lamchau,

Nice one! 

I noticed that the session.py file is getting bigger.  Just wondering whether it is better to separate some of these logics to another component or file?

lifeizhou-ap commented:
also I feel that with your change, we may not require `session resume` as user can `resume` the session via `session start`. WDYT?

lamchau commented:
@lifeizhou-ap maybe after we get to 500 lines or so unless you have another suggestion? 

30% or so of the `Session` class is text strings (split lines) and docstrings - with code folding it feels digestible. i'm worried if we start to make it to abstract it requires more context switching/modification across files

lamchau commented:
@lifeizhou-ap we could but to me if we got rid of it would take extra interactions/confirmations and keystrokes to get to resume (i.e. one more step to be an error). `start` doesn't sound like `resume`, but i'm biased so we should what flow makes the most sense. since it's a product call, maybe anna wants to chime in?

having a convenience method doesn't seem too bad and it's more or less "free"

lifeizhou-ap commented:
> @lifeizhou-ap maybe after we get to 500 lines or so unless you have another suggestion?
> 
> 30% or so of the `Session` class is text strings (split lines) and docstrings - with code folding it feels digestible. i'm worried if we start to make it to abstract it requires more context switching/modification across files

I see. Personally I prefer small files for readability and maintenance. However, I understand your concern and we can refactor in the future when it reaches the point for refactoring.

One suggestion: Maybe extract the code into a function? 
```
# prevents cluttering the `sessions` with empty files, which
        # can be confusing when resuming a session
        if is_empty_session(self.session_file_path):
            try:
                self.session_file_path.unlink()
            except FileNotFoundError:
                pass
            except Exception as e:
                raise Exception(f"error deleting empty session file: {e}")
```

lifeizhou-ap commented:
Hey @baxen I am happy with this PR (only with a small suggestion). Would you like to review it?

lamchau commented:
@lifeizhou-ap sure thing, extracted and reordered them here a1c8c82

lamchau commented on src/goose/cli/session.py:
```diff
@@ -140,13 +140,45 @@ def process_first_message(self) -> Optional[Message]:
 
+    def prompt_overwrite_session(self) -> None:
+        print(f"[yellow]session already exists at {self.session_file_path}.[/]\n")
+        print("would like to overwrite the existing session?")
```
renamed this function to `_prompt_overwrite_session` in a1c8c82

baxen commented on src/goose/cli/session.py:
```diff
@@ -140,13 +140,45 @@ def process_first_message(self) -> Optional[Message]:
+            self.prompt_overwrite_session()
+
+        print(
+            f"[bold]starting session | name: [cyan]{self.name}[/]  profile: [cyan]{self.profile or 'default'}[/][/bold]"
```
keep it dim imo! this info shows every time in case you need it but you mostly don't, so i like it being de-emphasized

baxen commented on src/goose/cli/session.py:
```diff
@@ -239,6 +245,55 @@ def _log_cost(self) -> None:
+    def _prompt_overwrite_session(self) -> None:
+        print(f"[yellow]session already exists at {self.session_file_path}.[/]\n")
+        print("would like to overwrite the existing session?")
+        while True:
```
nit: rich prompt encapsulates encapsulates the retry/invalid choice handling pretty well for this

lamchau replied:
ahh, good catch! glossed over that detail in the docs

lamchau replied:
[aa289ca](https://github.com/block-open-source/goose/pull/101/commits/aa289ca876b1944f501f408d187806c6c656f72b)

baxen reviewed (approved):
LGTM

lamchau commented:
> keep it dim imo! this info shows every time in case you need it but you mostly don't, so i like it being de-emphasized

@baxen it's pretty hard to see dim magenta, it'd be friendlier imo if we either keep it the same gray color and not emphasize it with color at all. or alternatively we don't display the file location at all?

![image](https://github.com/user-attachments/assets/2759cb5e-15d7-4066-a286-0a68bfe88ce6)

baxen commented:
> > keep it dim imo! this info shows every time in case you need it but you mostly don't, so i like it being de-emphasized
> 
> @baxen it's pretty hard to see dim magenta, it'd be friendlier imo if we either keep it the same gray color and not emphasize it with color at all. or alternatively we don't display the file location at all?
> 
> ![image](https://private-user-images.githubusercontent.com/4911400/373778763-2759cb5e-15d7-4066-a286-0a68bfe88ce6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjg0MDkzOTQsIm5iZiI6MTcyODQwOTA5NCwicGF0aCI6Ii80OTExNDAwLzM3Mzc3ODc2My0yNzU5Y2I1ZS0xNWQ3LTQwNjYtYTI4Ni0wYTY4YmZlODhjZTYucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAwOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMDhUMTczODE0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2VhOTQyNjc3ZDMxOTY4MDM5NGUyY2FmZDk2ZTQ0YzAzYWNkYmQxZjU4Y2I3N2NhNjI0N2Q2NWIxZjc4ZTc2MSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.seMTtbI3d2mj301If2MItl7TuXbWUUXJHhUydxL_U_I)

+1 to not needing tho show the file location at all here

lamchau commented:
@baxen for posterity
 - reverted dim [a1d4bba](https://github.com/block-open-source/goose/pull/101/commits/a1d4bbad87f3634be0151c81eec976d86a118be3)
 - removed session location [c3da1a4](https://github.com/block-open-source/goose/pull/101/commits/c3da1a448ff449aaf237fb0b8b7133e2170edba5)
